File: examples/cpp-examples/MeshOffset.dox.cpp
#include <MRMesh/MRBox.h>
#include <MRMesh/MRMesh.h>
#include <MRMesh/MRCube.h>
#include <MRMesh/MRMeshSave.h>
#include <MRMesh/MRRegionBoundary.h>
#include <MRVoxels/MROffset.h>

#include <iostream>

int main()
{
    // Create mesh
    MR::Mesh mesh = MR::makeCube();

//! [0]    
    // Setup parameters
    MR::GeneralOffsetParameters params;
    // calculate voxel size depending on desired accuracy and/or memory consumption
    params.voxelSize = suggestVoxelSize( mesh, 10000000.f );
    if ( !MR::findRightBoundary( mesh.topology ).empty() )
        params.signDetectionMode = MR::SignDetectionMode::HoleWindingRule; // use if you have holes in mesh

    // Make offset mesh
    float offset = mesh.computeBoundingBox().diagonal() * 0.1f;
    auto meshRes = MR::generalOffsetMesh( mesh, offset, params );
//! [0]    
    if ( !meshRes.has_value() )
    {
        std::cerr << meshRes.error() << std::endl;
        return 1;
    }

    // Save result
    if ( auto saveRes = MR::MeshSave::toAnySupportedFormat( *meshRes, "mesh_offset.stl" ); !saveRes )
    {
        std::cerr << saveRes.error() << std::endl;
        return 1;
    }

    return 0;
}

File: examples/js-examples/MeshOffset.dox.js
import createMeshLib from '@meshinspector/meshlib';

const ml = await createMeshLib();

// Create mesh
using mesh = ml.makeCube({ x: 1, y: 1, z: 1 }, { x: -0.5, y: -0.5, z: -0.5 });

// Setup parameters
using params = new ml.GeneralOffsetParameters();
// calculate voxel size depending on desired accuracy and/or memory consumption
params.voxelSize = ml.suggestVoxelSize(mesh, 10000000.0);
using topology = mesh.topology;
if (ml.findRightBoundary(topology).length > 0)
  params.signDetectionMode = ml.SignDetectionMode.HoleWindingRule; // use if you have holes in mesh

// Make offset mesh
using box = mesh.computeBoundingBox();
const offset = box.diagonal() * 0.1;
using result = ml.generalOffsetMesh(mesh, offset, params);

// Save result
ml.MeshSave.toAnySupportedFormat(result, 'mesh_offset.stl');
